master/internal/api_runs.go

```diff
@@ -403,3 +412,99 @@ func (a *apiServer) MoveRuns(
 	}
 	return &apiv1.MoveRunsResponse{Results: results}, nil
 }
+
+func (a *apiServer) KillRuns(ctx context.Context, req *apiv1.KillRunsRequest,
+) (*apiv1.KillRunsResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var runChecks []archiveKillRunOKResult
+	getQ := db.Bun().NewSelect().
+		ModelTableExpr("runs AS r").
+		Model(&runChecks).
+		Column("r.id").
+		ColumnExpr("COALESCE((e.archived OR p.archived OR w.archived), FALSE) AS archived").
+		ColumnExpr("t.request_id").
+		ColumnExpr("r.state IN (?) AS state", bun.In(model.StatesToStrings(model.TerminalStates))).
+		Join("LEFT JOIN trials_v2 t ON r.id=t.run_id").
+		Join("LEFT JOIN experiments e ON r.experiment_id=e.id").
+		Join("JOIN projects p ON r.project_id = p.id").
+		Join("JOIN workspaces w ON p.workspace_id = w.id").
+		Where("r.project_id = ?", req.ProjectId)
+
+	if req.Filter == nil {
+		getQ = getQ.Where("r.id IN (?)", bun.In(req.RunIds))
+	} else {
+		getQ, err = filterRunQuery(getQ, req.Filter)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if getQ, err = experiment.AuthZProvider.Get().
+		FilterExperimentsQuery(ctx, *curUser, nil, getQ,
+			[]rbacv1.PermissionType{rbacv1.PermissionType_PERMISSION_TYPE_UPDATE_EXPERIMENT}); err != nil {
+		return nil, err
+	}
+
+	err = getQ.Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []*apiv1.RunActionResult
+	var visibleIDs []int32
+	var validIDs []int32
+	for _, check := range runChecks {
+		visibleIDs = append(visibleIDs, check.ID)
+		switch {
+		case check.State:
+			results = append(results, &apiv1.RunActionResult{
+				Error: "",
+				Id:    check.ID,
+			})
+		case check.Archived:
+			results = append(results, &apiv1.RunActionResult{
+				Error: "Run is archived.",
+				Id:    check.ID,
+			})
```

Comment:
I think this case is sort of nonsense. Are we going to let people archive a run during execution? If we do and a run is archived but still running, should you really not be allowed to kill it? I think it should still be killable.

Reply:
Removed